bgpd/bgp_nhg.c

```diff
@@ -750,6 +750,101 @@ static void bgp_nhg_detach_nexthop(struct bgp_nhg_cache *nhg)
 		bgp_nhg_del_nhg(nhg);
 }
 
+/* called whenever a nhg is resolved over ourselves
+ * nexthop_active() detects compares prefix to install with the resolved route
+ * but with nhg, the prefix is not available, and we lose the loop detection
+ * this code replaces nhgs with regular route_add/delete to avoid this corner case
+ * for each path to install, use route_add/delete when route resolved matches the prefix
+ * for each path, we assume dependent paths are active, and are pushed again to zebra
+ */
+static bool bgp_nhg_try_detach_if_same_prefix(struct bgp_nhg_cache *nhg,
+					      struct prefix *resolved_prefix)
+{
+	struct bgp_nhg_connected *rb_node_dep = NULL;
+	struct bgp_nhg_cache *dep_nhg;
+	struct bgp_path_info *path, *safe;
+	const struct prefix *p;
+
+	frr_each_safe (bgp_nhg_connected_tree, &nhg->nhg_dependents_groups,
+		       rb_node_dep) {
+		dep_nhg = rb_node_dep->nhg;
+		LIST_FOREACH_SAFE (path, &(dep_nhg->paths), nhg_cache_thread,
+				   safe) {
+			p = bgp_dest_get_prefix(path->net);
+			if (path->bgp_nhg_nexthop == nhg &&
```

Comment:
Can you do reverse checks and `continue` to avoid this nesting level?

Reply:
of course